File: examples/python/grib_set_keys.py
# Copyright 2005-2018 ECMWF.
#
# This software is licensed under the terms of the Apache Licence Version 2.0
# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
#
# In applying this licence, ECMWF does not waive the privileges and immunities
# granted to it by virtue of its status as an intergovernmental organisation
# nor does it submit to any jurisdiction.

from __future__ import print_function
import traceback
import sys

from eccodes import *
from datetime import date
from collections import OrderedDict

INPUT = '../../data/regular_latlon_surface_constant.grib1'
OUTPUT = 'out.set.grib'
VERBOSE = 1  # verbose error reporting


def example():
    fin = open(INPUT)
    fout = open(OUTPUT, 'w')
    gid = codes_grib_new_from_file(fin)

    dt = date.today()
    today = "%d%02d%02d" % (dt.year, dt.month, dt.day)
    codes_set(gid, 'dataDate', int(today))
    codes_set(gid, 'centre', 80)

    centreIntVal = codes_get_array(gid, 'centre', int)
    centreStrVal = codes_get_array(gid, 'centre', str)
    dateStrVal = codes_get_array(gid, 'dataDate', str)
    assert (centreIntVal[0] == 80)
    assert (centreStrVal[0] == 'cnmc')
    assert (dateStrVal[0] == today)
    print('get centre as an integer - centre = %d' % centreIntVal[0])
    print('get centre as a string - centre = %s' % centreStrVal[0])
    print('get date as a string - date = %s' % dateStrVal[0])

    # Now do the same but using set_key_vals, setting keys all at once
    print('set keys using one long comma-separated string...')
    codes_set_key_vals(gid, 'level=1,centre=98')
    assert (codes_get(gid, 'centre', str) == 'ecmf')
    assert (codes_get(gid, 'level', int) == 1)

    print('set keys using a list of strings...')
    codes_set_key_vals(gid, ['level=2', 'centre=kwbc'])
    assert (codes_get(gid, 'centre', int) == 7)
    assert (codes_get(gid, 'level', int) == 2)

    print('set keys using a dictionary (order not as specified!)...')
    codes_set_key_vals(gid, {'level': 3, 'centre': 84})
    assert (codes_get(gid, 'centre', str) == 'lfpw')
    assert (codes_get(gid, 'level', int) == 3)

    print('set keys using an ordered dictionary...')
    codes_set_key_vals(gid, OrderedDict([('level', 3), ('centre', 84)]))
    assert (codes_get(gid, 'centre', str) == 'lfpw')
    assert (codes_get(gid, 'level', int) == 3)

    codes_gts_header(True)
    codes_gts_header(False)

    codes_write(gid, fout)
    codes_release(gid)
    fin.close()
    fout.close()


def main():
    try:
        example()
    except CodesInternalError as err:
        if VERBOSE:
            traceback.print_exc(file=sys.stderr)
        else:
            sys.stderr.write(err.msg + '\n')

        return 1


if __name__ == "__main__":
    sys.exit(main())

File: examples/C/grib_set_keys.c
/*
 * Copyright 2005-2018 ECMWF.
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
 * virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
 */

/*
 * C Implementation: grib_set_keys
 *
 * Description: how to set key values in GRIB messages
 *
 */
#include <stdio.h>
#include <stdlib.h>

#include "eccodes.h"

int main(int argc, char** argv)
{
    int err = 0;
    long centre=80;
    long long_value=0;
    char string_value[100];
    size_t len = sizeof(string_value)/sizeof(char);
    size_t size=0;

    FILE* in = NULL;
    const char* infile = "../../data/regular_latlon_surface.grib1";
    FILE* out = NULL;
    const char* outfile = "out.set.grib1";
    codes_handle *h = NULL;
    const void* buffer = NULL;
    size_t str_len = 0; /* See the call to codes_set_string later */

    in = fopen(infile,"r");
    if(!in) {
        printf("ERROR: unable to open file %s\n",infile);
        return 1;
    }

    out = fopen(outfile,"w");
    if(!out) {
        printf("ERROR: unable to open file %s\n",outfile);
        fclose(in);
        return 1;
    }

    /* create a new handle from a message in a file */
    h = codes_handle_new_from_file(0, in, PRODUCT_GRIB, &err);
    if (h == NULL) {
        printf("Error: unable to create handle from file %s\n",infile);
    }

    /* set centre as a long */
    CODES_CHECK(codes_set_long(h,"centre",centre),0);

    /* set paramId and shortName - normally you would do one or the other */
    CODES_CHECK(codes_set_long(h,"paramId", 500004),0);
    /* the value of str_len is not used, it can be anything! */
    CODES_CHECK(codes_set_string(h,"shortName", "fis", &str_len),0);

    /* get centre as a long */
    CODES_CHECK(codes_get_long(h,"centre",&long_value),0);
    printf("centre long value=%ld\n",long_value);

    /* get centre as a string */
    CODES_CHECK(codes_get_string(h,"centre",string_value,&len),0);
    printf("centre string value=%s\n",string_value);

    /* get the coded message in a buffer */
    CODES_CHECK(codes_get_message(h,&buffer,&size),0);

    /* write the buffer in a file*/
    if(fwrite(buffer,1,size,out) != size)
    {
        perror(argv[1]);
        exit(1);
    }

    codes_handle_delete(h);
    fclose(in);
    fclose(out);

    return 0;
}
